### DISCUSSION TITLE
Question about punching and bending

### DISCUSSION THREAD

**Discussion by WhereMyRPI** [CadQuery/cadquery#1193] [Q&A]
Hello, 

I'm stuck trying to automatically create busbars that are firstly **punched** and secondly **bended**.

Creating such a busbar with the appropriate punchings is easy, as one of the examples in the documentation [(link)](https://cadquery.readthedocs.io/en/latest/examples.html#making-counter-bored-and-counter-sunk-holes) shows:

![image](https://user-images.githubusercontent.com/117776938/200671684-eca2ac9d-189b-4914-b281-018a4df91bfd.png)

This punching should be done first, because the coordinates of the punchings that are available are expressed in terms of the flat busbar.

The next step is to bend this busbar. The result should look something like this:

![image](https://user-images.githubusercontent.com/117776938/200672618-b4107af7-2f10-456f-ba7d-6a733925a7d6.png)

My problem is that is can't figure out how to do this bending. 

I could bend the busbar first by creating and extruding a sketch, but placing the punchings would be very difficult with this approach, right?

Best regards.

**Comment by lorenzncode** [comment]
Here is an experiment to start the discussion.  Perhaps a higher level procedure could be built around something like this to automatically rotate the faces as in the following.


```py
s = cq.Sketch().rect(20, 60).push([(0, 24), (0, -24)]).circle(2, mode="s")

f = s.reset().faces().val()

f0 = f.split(cq.Face.makePlane(None, None, cq.Vector(0, 10, 0), cq.Vector(0, -1, 0)))

w0 = (
    cq.Workplane(f0)
    .faces("<<Y")
    .tag("f1")
    .end(1)
    .faces(">>Y")
    .tag("f2")  # keep
    .faces(tag="f1")
    .rotate(cq.Vector(0, 10, 0), cq.Vector(1, 10, 0), 90)
    .faces()
    .tag("f3")
)

f1 = w0.faces(tag="f1")
f2 = w0.faces(tag="f2")
f3 = w0.faces(tag="f3")

f4 = w0.val().split(
    cq.Face.makePlane(None, None, cq.Vector(0, 0, -20), cq.Vector(0, 0, 1))
)

w1 = (
    cq.Workplane(f4)
    .faces("<<Z")
    .tag("f5")
    .end(1)
    .faces(">>Z")
    .tag("f6")  # keep
    .faces(tag="f5")
    .rotate(cq.Vector(0, 10, -20), cq.Vector(1, 10, -20), -90)
    .faces()
    .tag("f7")  # keep
)


f5 = w1.faces(tag="f5")
f6 = w1.faces(tag="f6")
f7 = w1.faces(tag="f7")

face = cq.Face(f2.val().fuse(f6.val(), f7.val()).wrapped)
result = face.thicken(2).fuse(face.thicken(-2))
```

**Reply by WhereMyRPI** [reply]
Hi, wow and thanks for the quick reply. That's a really interesting approach, I'll try to fiddle with that. There is no bend operation or something similar available in cadquery, is that correct?

Best regards

**Reply by lorenzncode** [reply]
Thanks for opening the discussion!  That's correct - there is no bend operation.  split/rotate was an attempt to create a similar shape to a bend.
